crates/cargo-test-macro/src/lib.rs

```diff
@@ -144,18 +147,44 @@ pub fn cargo_test(attr: TokenStream, item: TokenStream) -> TokenStream {
                     continue;
                 }
             }
+            TokenTree::Ident(i) => {
+                // The first time through it will be `fn` the second time is the
+                // name of the test.
+                if test_name.is_none() && num == 1 {
+                    test_name = Some(i.to_string())
+                } else {
+                    num += 1;
+                }
+                ret.extend(Some(TokenTree::Ident(i)));
+                continue;
+            }
             other => {
                 ret.extend(Some(other));
                 continue;
             }
         };
+        std::file!();
+        let name = test_name
+            .clone()
+            .map(|n| n.split("::").next().unwrap().to_string())
+            .unwrap();
 
-        let mut new_body = to_token_stream(
-            r#"let _test_guard = {
+        let mut new_body = to_token_stream(&format!(
+            r#"let _test_guard = {{
                 let tmp_dir = option_env!("CARGO_TARGET_TMPDIR");
-                cargo_test_support::paths::init_root(tmp_dir)
-            };"#,
-        );
+                let file_name = std::path::PathBuf::from(
```

Comment:
Could we add a simple comment here so the desired output is more clear without tracing the whole source? Such as

```markdown
test fn `workspaces::workspace_in_git` 
  -> `file!` expands to `tests/testsuite/workspaces.rs`
  -> `workspaces`
test fn `cargo_add::add_basic::add_basic`
  -> `file!` expands to `tests/testsuite/cargo_add/add_basic.rs`
  -> `cargo_add`
```